### PR TITLE
Storage dictionary compression

### PR DESCRIPTION
**Pull request by samansmink**
This PR adds dictionary compression for strings to the DuckDB storage format. Dictionary compression uses a combination of bitpacking and a dictionary to compress string segments. The data is stored across three buffers: the index buffer, the selection buffer and the dictionary. Firstly the Index buffer contains the offsets into the dictionary which are also used to determine the string lengths. Each value in the dictionary gets a single unique index in the index buffer. Secondly, the selection buffer maps the tuples to an index in the index buffer. The selection buffer is compressed with bitpacking. Finally, the dictionary contains simply all the unique strings without lengths or null termination as we can deduce the lengths from the index buffer. The addition of the selection buffer is done for two reasons: firstly, to allow the scan to emit dictionary vectors by scanning the whole dictionary at once and then scanning the selection buffer for each emitted vector. Secondly, it allows for efficient bitpacking compression as the selection values should remain relatively small.

# Evaluation
For evaluation I've also included my initial version of dictionary compression which is simply the current string_uncompressed.cpp with a dictionary to deduplicate strings. This implementation is called duplicate elimination but is not included in this PR as the current implementation outperforms it. Especially 

## TPC-H storage size
compression | size
-- | --
Uncompressed | 1.1G
Bitpacking + RLE | 855M
DuckDB master + Duplicate elimination | 650M
DuckDB master + Dictionary Compression | 540M

## Microbenchmarks
Firstly some micro benchmarks, these are added in the pr in benchmark/micro/compression/dictionary/*

benchmark | Uncompressed | Dictionary Compression | Duplicate elimination | dictionary + bitpacking relative | duplicate elimination relative
--  | -- | -- | -- | -- | --
read (~4x compress ratio) | 0.61251750 | 0.39860225 | 0.60771200 | -35% | -1%
read ("max" compress ratio) | 0.60801640 | 0.38863540 | 0.60300940 | -36% | -1%
read (uncompressible data) | 0.92532780 | 1.01706320 | 0.92650340 | 10% | 0%
write (~4x compress ratio) | 15.02338560 | 17.88984900 | 22.16934540 | 19% | 48%
write (uncompressible data) | 9.11934460 | 17.34924600 | 21.04727860 | 90% | 131%

## TPCH SF10
Query performance overall improves with this implementation. Not only is there less data to scan, but emitting dictionary vectors can be done quite efficiently. Note that I compare the current implementation to the current master, where RLE and bitpacking compression are enabled.

benchmark | no varchar compression | dictionary compression | relative
 --  | -- | -- | --
benchmark/tpch/sf1/q01.benchmark | 0.52060175 | 0.39091380 | -25%
benchmark/tpch/sf1/q02.benchmark | 0.08866200 | 0.08695160 | -2%
benchmark/tpch/sf1/q03.benchmark | 0.43471620 | 0.43098680 | -1%
benchmark/tpch/sf1/q04.benchmark | 0.88235000 | 0.85945580 | -3%
benchmark/tpch/sf1/q05.benchmark | 0.26422160 | 0.26628160 | 1%
benchmark/tpch/sf1/q06.benchmark | 0.08281600 | 0.08265860 | -0%
benchmark/tpch/sf1/q07.benchmark | 1.00017580 | 0.97276060 | -3%
benchmark/tpch/sf1/q08.benchmark | 0.28746000 | 0.28878060 | 0%
benchmark/tpch/sf1/q09.benchmark | 4.70156680 | 4.71601820 | 0%
benchmark/tpch/sf1/q10.benchmark | 0.45207260 | 0.42123720 | -7%
benchmark/tpch/sf1/q11.benchmark | 0.05219020 | 0.05422400 | 4%
benchmark/tpch/sf1/q12.benchmark | 0.78107160 | 0.65350440 | -16%
benchmark/tpch/sf1/q13.benchmark | 0.64187040 | 0.68768860 | 7%
benchmark/tpch/sf1/q14.benchmark | 0.14572480 | 0.14204740 | -3%
benchmark/tpch/sf1/q15.benchmark | 0.34207700 | 0.34491580 | 1%
benchmark/tpch/sf1/q16.benchmark | 1.28658660 | 1.22756540 | -5%
benchmark/tpch/sf1/q17.benchmark | 1.63868940 | 1.63657780 | -0%
benchmark/tpch/sf1/q18.benchmark | 1.60507840 | 1.59268660 | -1%
benchmark/tpch/sf1/q19.benchmark | 0.44794620 | 0.32764120 | -27%
benchmark/tpch/sf1/q20.benchmark | 0.61579500 | 0.61497460 | -0%
benchmark/tpch/sf1/q21.benchmark | 1.88817040 | 1.88623420 | -0%
benchmark/tpch/sf1/q22.benchmark | 0.47655720 | 0.47545760 | -0%
sum | 18.63639995 | 18.15956240 | -3%
avg | 0.84710909 | 0.82543465 | -3%

compression | size
--- | ---
no varchar compression | 9,02GB
dictionary compression | 5,77GB

# Next steps
- Currently there is no patching/exception mechanism this should not be too difficult to implement next.
- Currently the null value is always added to the dict as it simplifies the code a bit. Removing it if we don't have will slightly improve compression. Should be easy to implement together with patching
- Dictionary compression support for fixed-size datatypes

### CODE REVIEW

File: src/storage/compression/dictionary_compression.cpp [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+	DictionaryCompressionAnalyzeState()
+	    : segment_count(0), current_tuple_count(0), current_unique_count(0), current_dict_size(0) {
+		current_set = make_unique<std::unordered_set<string>>();
+	}
+
+	size_t segment_count;
+	idx_t current_tuple_count;
+	idx_t current_unique_count;
+	size_t current_dict_size;
+	unique_ptr<std::unordered_set<string>> current_set;
```
Comment (Mytherin) [inline src/storage/compression/dictionary_compression.cpp:314]: Does this need to be a unique_ptr?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great. Impressive results! Some comments:

**Comment by Mytherin** [comment]
There also seems to be a test failure in the vector sizes still:

```
[1192/1470]: test/sql/storage/compression/dictionary/dictionary_index_fetch.test
[2150](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2150)
FAILURE IN RUNNING TEST
[2151](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2151)
--------------------
[2152](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2152)
RETURNCODE
[2153](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2153)
--------------------
[2154](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2154)

[2155](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2155)
1
[2156](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2156)
--------------------
[2157](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2157)
STDOUT
[2158](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2158)
--------------------
[2159](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2159)

[2160](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2160)
Filters: test/sql/storage/compression/dictionary/dictionary_index_fetch.test
[2161](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2161)

[2162](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2162)
[0/1] (0%): test/sql/storage/compression/dictionary/dictionary_index_fetch.test 
[2163](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2163)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2164](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2164)
unittest is a Catch v2.13.7 host application.
[2165](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2165)
Run with -? for options
[2166](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2166)

[2167](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2167)
-------------------------------------------------------------------------------
[2168](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2168)
test/sql/storage/compression/dictionary/dictionary_index_fetch.test
[2169](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2169)
-------------------------------------------------------------------------------
[2170](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2170)
/home/runner/work/duckdb/duckdb/test/sqlite/test_sqllogictest.cpp:148
[2171](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2171)
...............................................................................
[2172](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2172)

[2173](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2173)
test/sql/storage/compression/dictionary/dictionary_index_fetch.test:25: FAILED:
[2174](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2174)
explicitly with message:
[2175](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2175)
  0
[2176](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2176)

[2177](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2177)

[2178](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2178)
[1/1] (100%): test/sql/storage/compression/dictionary/dictionary_index_fetch....
[2179](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2179)
===============================================================================
[2180](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2180)
test cases: 1 | 1 failed
[2181](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2181)
assertions: 6 | 5 passed | 1 failed
[2182](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2182)

[2183](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2183)

[2184](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2184)
--------------------
[2185](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2185)
STDERR
[2186](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2186)
--------------------
[2187](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2187)

[2188](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2188)
================================================================================
[2189](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2189)
Query unexpectedly failed (test/sql/storage/compression/dictionary/dictionary_index_fetch.test:25)
[2190](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2190)
================================================================================
[2191](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2191)
SQL Query
[2192](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2192)
SELECT MIN(id), MAX(id), SUM(col::INT), MIN(col::INT), MAX(col::INT), COUNT(*) FROM test WHERE id=5000
[2193](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2193)
================================================================================
[2194](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2194)
Actual result:
[2195](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2195)
INTERNAL Error: INTERNAL Error: Assertion triggered in file "/home/runner/work/duckdb/duckdb/src/include/duckdb/storage/table/column_segment.hpp" on line 104: row_index >= this->start
[2196](https://github.com/duckdb/duckdb/runs/5251441433?check_suite_focus=true#step:4:2196)`
```

File: benchmark/micro/compression/dictionary/dictionary_tpch_large_write.benchmark [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+name Bitpacking Insert
+group aggregate
+storage persistent
+require_reinit
+
+require tpch
+
+load
+PRAGMA force_compression='uncompressed';
```
[the remark this answers is not in the document]
Comment (samansmink) [reply benchmark/micro/compression/dictionary/dictionary_tpch_large_write.benchmark]: this file shouldnt have been committed

**Comment by hawkfish** [comment]
Very exciting!

Do you know why Q13 got so much slower? It seems to be reading a comment field, which probably doesn't compress.  I'm also puzzled by q11 is slower - it is just reading a small cardinality string field.

Are we planning to add [dictionary joins](https://research.tableau.com/sites/default/files/SIGMOD%202014.pdf) against the dictionary compressed columns? For string predicatees that can be very beneficial.

**Comment by samansmink** [comment]
@hawkfish Q13 is indeed slower because of the poorly compressible comment field, it compresses just a tiny bit (compression ratio is 1.002), making duckdb choose dictionary compression. As shown by the microbencharks, the implementation of dictionary compression on poorly compressible data results in a slight read overhead. I think we may want to add a threshold compression ratio for duckdb when selecting the compression method. 

I agree Q11 shouldn't be slower. I think the results might be noise, i'll see if i can reproduce the result.

**Comment by Mytherin** [comment]
Could you have a look at the failing test when you have time? It seems related to this PR.

```
[162/1724] (9%): Test scanning a table and computing an aggregate over a tabl...
[589](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:589)
Query failed with message: TransactionContext Error: Failed to commit: INTERNAL Error: Invalid unicode detected in segment statistics update!
[590](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:590)
[163/1724] (9%): Test storing a big string that exceeds buffer manager size     
[591](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:591)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[592](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:592)
unittest is a Catch v2.13.7 host application.
[593](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:593)
Run with -? for options
[594](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:594)

[595](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:595)
-------------------------------------------------------------------------------
[596](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:596)
Test storing a big string that exceeds buffer manager size
[597](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:597)
-------------------------------------------------------------------------------
[598](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:598)
/__w/duckdb/duckdb/test/sql/storage/test_buffer_manager.cpp:54
[599](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:599)
...............................................................................
[600](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:600)

[601](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:601)
/__w/duckdb/duckdb/test/sql/storage/test_buffer_manager.cpp:74: FAILED:
[602](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:602)
  REQUIRE( NO_FAIL((con.Query("INSERT INTO test SELECT a||a||a||a||a||a||a||a||a||a, " + to_string(iteration) + " FROM test"))) )
[603](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:603)
with expansion:
[604](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:604)
  false
[605](https://github.com/duckdb/duckdb/runs/5272508404?check_suite_focus=true#step:9:605)
```

**Comment by Mytherin** [comment]
Looks excellent, thanks for all the updates!

**Pull request merged**